feat(node): Expose functional integrations to replace classes

mydea opened the pull request:
This updates the general (non-tracing) node integrations to be functional.

For node & undici, the replacements are slightly different:

* `new Http()` --> `httpIntegration()`: In contrast to the class integration, this will create spans by default if tracing is enabled. While at it, this also "fixes" that if `tracing: false` is set, no spans will be created.
* `new Undici()` --> `nativeNodeFetchIntegration()`: Renamed this for consistency, and added a `tracing` option similar to http to allow to disable span creation.

We can't really deprecate `Integrations.xxx` yet until we have replacements for the tracing integrations 😬 So that would also be a todo left.

AbhiPrasad commented on packages/node/src/integrations/undici/index.ts:
```diff
@@ -36,6 +37,13 @@ export enum ChannelName {
+  return new Undici(options) as unknown as IntegrationFnResult;
+}) satisfies IntegrationFn;
+
+export const undiciIntegration = defineIntegration(_undiciIntegration);
```
we should probably rename this right now tbh. Nobody recognizes Undici 😢

mydea replied:
ahh right! Let's go with `nativeNodeFetchIntegration()` as Luca proposed some time ago 👍

mydea replied:
Renamed it!

mydea commented:
Update: So I updated the node http/fetch integrations so that the functional integrations have slightly new/improved behavior:

With them, span creation will be automatically enabled if tracing is enabled - so no more need to add `httpIntergation({ tracing: true })`. You can still set it to `false` but the default is different. I also added some tests to cover this.

Lms24 commented on packages/node/src/integrations/local-variables/local-variables-async.ts:
Noaction required/General comment: Are we actually exporting the async localvariables integration? Already stumbled upon this yesterday and wasn't sure.

mydea replied:
no, just the sync one right now!

Lms24 commented on packages/node/src/integrations/undici/index.ts:
```diff
@@ -44,6 +46,13 @@ export interface UndiciOptions {
    */
   breadcrumbs: boolean;
+
+  /**
+   * Whether tracing spans should be created for requests
+   * If not set, this will be enabled/disabled based on if tracing is enabled.
+   */
+  tracing?: boolean;
```
l: I'd slightly reword this based on my understanding that this option not only controls span creation but also distributed tracing:


```suggestion
  /**
   * Controls whether tracing spans should be created for requests and if 
   * distributed tracing headers should be attached to outgoing requests.
   *
   * @default If not set, tracing follows global SDK settings (`tracesSampleRate`).
   */
  tracing?: boolean;
```

Lms24 commented on MIGRATION.md:
```diff
@@ -34,34 +34,45 @@ integrations from the `Integrations.XXX` hash, is deprecated in favor of using t
+| `new DenoContext()`          | `denoContextIntegration()`          | `@sentry/deno`                                                                                          |
+| `new DenoCron()`             | `denoCronIntegration()`             | `@sentry/deno`                                                                                          |
+| `new NormalizePaths()`       | `normalizePathsIntegration()`       | `@sentry/deno`                                                                                          |
+| `new Console()`              | `consoleIntegration()`              | `@sentry/node`                                                                                          |
+| `new Context()`              | `nodeContextIntegration()`          | `@sentry/node`                                                                                          |
+| `new Modules()`              | `modulesIntegration()`              | `@sentry/node`                                                                                          |
+| `new OnUncaughtException()`  | `onUncaughtExceptionIntegration()`  | `@sentry/node`                                                                                          |
+| `new OnUnhandledRejection()` | `onUnhandledRejectionIntegration()` | `@sentry/node`                                                                                          |
+| `new LocalVariables()`       | `localVariablesIntegration()`       | `@sentry/node`                                                                                          |
+| `new Spotlight()`            | `spotlightIntergation()`            | `@sentry/node`                                                                                          |
+| `new Anr()`                  | `anrIntergation()`                  | `@sentry/node`                                                                                          |
+| `new Hapi()`                 | `hapiIntegration()`                 | `@sentry/node`                                                                                          |
+| `new Undici()`               | `nativeNodeFetchIntegration()`      | `@sentry/node`                                                                                          |
+| `new Http()`                 | `httpIntegration()`                 | `@sentry/node`                                                                                          |
```
I feel like we need to do a better job to call out that this also applies to all `@sentry/node`-based packages (serverless, fullstack framework SDKs). Doesn't strictly have to happen now but I'd say we should add them. Probably simplest to list them individually.

mydea replied:
yeah, I thought about this too but this makes the table veeery unwieldly 😓  not sure, maybe we just point it out above? E.g. node also applies to next, remix, sveltekit, ..., browser also applies to react, ...?
